From 905b715c5f5bdf4acfd7caa25637116904f9f3b7 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 02 Sep 2022 12:46:18 +0300
Subject: [PATCH] [SW-102303] gaudi3: update fake fuse indication to slv mechanism

Currently, to receive fake fuse data from master, slave die polls
on master die sw status bit whether master die completed fuse init.
It was recommendation by HW team to use slave die register to indicate
slave die above fake fuse data availability. This patch updates logic to
use scratchpad register of slave die to indicate fake fuse data
is available.

Change-Id: I9e5e754b026a6666b34e5997d19d38d128ef2231
---

diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index 13419cf..9836965 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -9,6 +9,9 @@
 #define GPIOS_PER_BANK			16
 #define TPM_CS_GPIO_PIN			18
 
+/* Fuse emulation indication magic:	  F I N D (Fuse INDication) */
+#define FFUSE_WRITE_IND_MAGIC		0x46494E44
+
 enum pll_div_out {
 	PLL_DIV0,
 	PLL_DIV1,
diff --git a/arch/gaudi3/arc/regmap.h b/arch/gaudi3/arc/regmap.h
index 5195d6e..83a994f 100644
--- a/arch/gaudi3/arc/regmap.h
+++ b/arch/gaudi3/arc/regmap.h
@@ -51,6 +51,7 @@
 #define mmSPI_DIRECT_WR_RD_CTRL		mmD0_PSOC_GLOBAL_CONF_SPI_DIRECT_WR_RD_CTRL
 #define mmSYSCFG1_EFUSE_DATA		mmD0_PSOC_GLOBAL_CONF_SCRATCHPAD_0
 #define mmSYSCFG2_EFUSE_DATA		mmD0_PSOC_GLOBAL_CONF_SCRATCHPAD_1
+#define mmFFUSE_WRITE_IND		mmD0_PSOC_GLOBAL_CONF_SCRATCHPAD_2
 
 #define mmGPIO0_DATA			mmD0_PSOC_GPIO0_DATA
 
diff --git a/arch/gaudi3/arc/slave.c b/arch/gaudi3/arc/slave.c
index 39dcfeb..79f2eab 100644
--- a/arch/gaudi3/arc/slave.c
+++ b/arch/gaudi3/arc/slave.c
@@ -45,9 +45,12 @@
 	uint32_t retry = FUSE_EMULATION_DATA_RETRIES;
 	uint32_t val;
 
+	/* Clear indication before polling */
+	regmap_writel(0x0, mmFFUSE_WRITE_IND);
+
 	while (1) {
-		val = mirror_readl(mmD0_PSOC_SECURITY_SW_STATUS_0);
-		if (val & BIT(PROGRESS_INIT_EFUSE_SPI))
+		val = regmap_readl(mmFFUSE_WRITE_IND);
+		if (val == FFUSE_WRITE_IND_MAGIC)
 			break;
 
 		retry--;
diff --git a/arch/gaudi3/arc/spi_flash.c b/arch/gaudi3/arc/spi_flash.c
index 7034f9b..4102dc3 100644
--- a/arch/gaudi3/arc/spi_flash.c
+++ b/arch/gaudi3/arc/spi_flash.c
@@ -71,6 +71,9 @@
 	/* Notify slave config fuse value in slave scratchpad 0/1 register */
 	mirror_writel(d1_syscfg1, (mmSYSCFG1_EFUSE_DATA + HL_CFG_SPACE_DIE1_OFFSET));
 	mirror_writel(d1_syscfg2, (mmSYSCFG2_EFUSE_DATA + HL_CFG_SPACE_DIE1_OFFSET));
+
+	/* Indicate slave die about fuse value is available */
+	mirror_writel(FFUSE_WRITE_IND_MAGIC, (mmFFUSE_WRITE_IND + HL_CFG_SPACE_DIE1_OFFSET));
 }
 
 void spi_flash_fuse_init(void)
